File: baseble/src/main/java/com/fcbox/ble/baseble/core/BluetoothGattChannel.java
package com.fcbox.ble.baseble.core;

import android.bluetooth.BluetoothGatt;
import android.bluetooth.BluetoothGattCharacteristic;
import android.bluetooth.BluetoothGattDescriptor;
import android.bluetooth.BluetoothGattService;


import com.fcbox.ble.baseble.common.PropertyType;

import java.util.UUID;

/**
 * BluetoothGatt 相关信息
 */
@SuppressWarnings("unused")
public class BluetoothGattChannel {

    private BluetoothGatt bluetoothGatt;
    private BluetoothGattService service;
    private BluetoothGattCharacteristic characteristic;
    private BluetoothGattDescriptor descriptor;
    private String gattInfoKey;
    private PropertyType propertyType;
    private UUID serviceUUID;
    private UUID characteristicUUID;
    private UUID descriptorUUID;

    private BluetoothGattChannel(BluetoothGatt bluetoothGatt, PropertyType propertyType, UUID serviceUUID, UUID characteristicUUID, UUID descriptorUUID) {
        this.bluetoothGatt = bluetoothGatt;
        this.propertyType = propertyType;
        this.serviceUUID = serviceUUID;
        this.characteristicUUID = characteristicUUID;
        this.descriptorUUID = descriptorUUID;
        StringBuilder stringBuilder = new StringBuilder();
        if (propertyType != null) {
            stringBuilder.append(propertyType.getPropertyValue());
        }
        if (serviceUUID != null && bluetoothGatt != null) {
            service = bluetoothGatt.getService(serviceUUID);
            stringBuilder.append(serviceUUID.toString());
        }
        if (service != null && characteristicUUID != null) {
            characteristic = service.getCharacteristic(characteristicUUID);
            stringBuilder.append(characteristicUUID.toString());
        }
        if (characteristic != null && descriptorUUID != null) {
            descriptor = characteristic.getDescriptor(descriptorUUID);
            stringBuilder.append(descriptorUUID.toString());
        }
        gattInfoKey = stringBuilder.toString();
    }

    public BluetoothGatt getBluetoothGatt() {
        return bluetoothGatt;
    }

     BluetoothGattCharacteristic getCharacteristic() {
        return characteristic;
    }

     BluetoothGattDescriptor getDescriptor() {
        return descriptor;
    }

    public BluetoothGattService getService() {
        return service;
    }

     BluetoothGattChannel setDescriptor(BluetoothGattDescriptor descriptor) {
        this.descriptor = descriptor;
        return this;
    }

     String getGattInfoKey() {
        return gattInfoKey;
    }

    public UUID getCharacteristicUUID() {
        return characteristicUUID;
    }

    public UUID getDescriptorUUID() {
        return descriptorUUID;
    }

     PropertyType getPropertyType() {
        return propertyType;
    }

    public UUID getServiceUUID() {
        return serviceUUID;
    }

    public static class Builder {
        private BluetoothGatt bluetoothGatt;
        private PropertyType propertyType;
        private UUID serviceUUID;
        private UUID characteristicUUID;
        private UUID descriptorUUID;

        public Builder() {
        }

        public Builder setBluetoothGatt(BluetoothGatt bluetoothGatt) {
            this.bluetoothGatt = bluetoothGatt;
            return this;
        }

        public Builder setCharacteristicUUID(UUID characteristicUUID) {
            this.characteristicUUID = characteristicUUID;
            return this;
        }

        public Builder setDescriptorUUID(UUID descriptorUUID) {
            this.descriptorUUID = descriptorUUID;
            return this;
        }

        public Builder setPropertyType(PropertyType propertyType) {
            this.propertyType = propertyType;
            return this;
        }

        public Builder setServiceUUID(UUID serviceUUID) {
            this.serviceUUID = serviceUUID;
            return this;
        }

        public BluetoothGattChannel builder() {
            return new BluetoothGattChannel(bluetoothGatt, propertyType, serviceUUID, characteristicUUID, descriptorUUID);
        }
    }
}

File: baseble/src/main/java/com/xzy/ble/baseble/core/BluetoothGattChannel.kt
package com.xzy.ble.baseble.core

import android.bluetooth.BluetoothGatt
import android.bluetooth.BluetoothGattCharacteristic
import android.bluetooth.BluetoothGattDescriptor
import android.bluetooth.BluetoothGattService
import com.xzy.ble.baseble.common.PropertyType


import java.util.UUID

/**
 * BluetoothGatt 相关信息
 */
@Suppress("unused")
class BluetoothGattChannel private constructor(
        bluetoothGatt: BluetoothGatt?, internal val propertyType: PropertyType?, serviceUUID: UUID?, characteristicUUID: UUID?, descriptorUUID: UUID?) {
    var service: BluetoothGattService? = null
        private set
    internal var characteristic: BluetoothGattCharacteristic? = null
        private set
    var descriptor: BluetoothGattDescriptor? = null
    val gattInfoKey: String

    init {
        val stringBuilder = StringBuilder()
        if (propertyType != null) {
            stringBuilder.append(propertyType.propertyValue)
        }
        if (serviceUUID != null && bluetoothGatt != null) {
            service = bluetoothGatt.getService(serviceUUID)
            stringBuilder.append(serviceUUID.toString())
        }
        if (service != null && characteristicUUID != null) {
            characteristic = service!!.getCharacteristic(characteristicUUID)
            stringBuilder.append(characteristicUUID.toString())
        }
        if (characteristic != null && descriptorUUID != null) {
            descriptor = characteristic!!.getDescriptor(descriptorUUID)
            stringBuilder.append(descriptorUUID.toString())
        }
        gattInfoKey = stringBuilder.toString()
    }

    internal fun getDescriptor(): BluetoothGattDescriptor? {
        return descriptor
    }

    internal fun setDescriptor(descriptor: BluetoothGattDescriptor): BluetoothGattChannel {
        this.descriptor = descriptor
        return this
    }

    class Builder {
        private var bluetoothGatt: BluetoothGatt? = null
        private var propertyType: PropertyType? = null
        private var serviceUUID: UUID? = null
        private var characteristicUUID: UUID? = null
        private var descriptorUUID: UUID? = null

        fun setBluetoothGatt(bluetoothGatt: BluetoothGatt): Builder {
            this.bluetoothGatt = bluetoothGatt
            return this
        }

        fun setCharacteristicUUID(characteristicUUID: UUID): Builder {
            this.characteristicUUID = characteristicUUID
            return this
        }

        fun setDescriptorUUID(descriptorUUID: UUID): Builder {
            this.descriptorUUID = descriptorUUID
            return this
        }

        fun setPropertyType(propertyType: PropertyType): Builder {
            this.propertyType = propertyType
            return this
        }

        fun setServiceUUID(serviceUUID: UUID): Builder {
            this.serviceUUID = serviceUUID
            return this
        }

        fun builder(): BluetoothGattChannel {
            return BluetoothGattChannel(bluetoothGatt, propertyType, serviceUUID, characteristicUUID, descriptorUUID)
        }
    }
}
